### DISCUSSION TITLE
Q&A on the Coreform Cubit Webinar

### DISCUSSION THREAD

**Discussion by aprilnovak** [neams-th-coe/cardinal#972] [Q&A]
The KRUSTY model from the Coreform Cubit webinar hosted on October 25 is freely available in the Cardinal repository. Please use this discussion page for questions related to running this model, compiling Cardinal, or Cubit usage.

**Comment by Kevrog1006** [comment]
Hello!

I attempted to use Cubit to create a model containing a helical fuel rod, cladding, and water. I also referred to the surface mesh generation steps of the KRUSTY model in the Cardinal tutorials named csg_to_cad. However, I encountered a problem when forming the block. The error message is as follows:

`Cubit>tetmesh tri all make block ERROR: surface is not closed or has dangling triangle(s) at edges: 8 9 7 1 10 5 6 11 12 13 2 3 4 35 47 56 64 51 69 63 16 41 30 45 52 57 58 67 37 18 71 23 50 33 70 73 75 39 49 61 62 17 46 55 14 26 65 76 66 77 29 40 48 36 21 25 31 32 24 22 42 20 53 54 59 34 60 15 68 72 74 19 27 28 38 43 44 94 95 104 112 102 92 123 80 83 97 111 114 116 98 119 100 82 113 87 101 121 128 130 132 133 93 134 96 81 107 108 90 109 117 131 135 136 137 140 141 99 85 89 126 138 88 86 103 84 115 120 106 124 139 79 110 105 129 118 91 122 125 127 78 166 157 175 142 180 168 146 159 143 173 161 182 148 149 164 145 156 158 177 178 152 187 188 155 193 196 150 153 162 147 176 185 189 197 199 201 195 203 204 205 144 171 183 186 198 184 169 202 160 165 190 151 194 192 181 174 200 179 163 167 170 172 191 154 206 231 232 239 208 237 222 233 213 218 228 216 227 230 226 217 236 219 238 224 225 235 210 240 209 215 229 234 221 214 211 207 212 223 220 ERROR: Command Failed.`
I'm not sure which step was wrong. Currently, I directly generated the volumetric mesh instead of combining and stitching surface meshes to form the volumetric mesh.Then, I exported it as a DAGMC file and an Exodus file: `Cubit>mesh volume 1 2 3
Matching intervals successful.
Begin tri meshing 39 surfaces on volumes 1 to 3
Completed tri meshing 39 surfaces with 781716 tris
Begin tet meshing 3 volumes
Completed meshing volume 1 with 2017679 tets
Mesh indicates volume 2 (Volume 2) has 1+ through holes.
Completed meshing volume 2 with 979616 tets
Mesh indicates volume 3 (Volume 3) has 1+ through holes.
Completed meshing volume 3 with https://github.com/neams-th-coe/cardinal/commit/13616050750ddc91c43ac26361fac49fd8edca9c tets
Volumes 1 to 3 meshing completed using scheme: tetmesh
Generated 4358900 tets
Meshing time: 109.906250

Finished Command: mesh volume 1 2 3`
Is this approach feasible?

Besides, I use OpenMC to set the boundary conditions, and the part surrounding the fuel rod adopts the total reflection condition. The code is as follows:
`dagmc_univ = openmc.DAGMCUniverse(filename="HCF.h5m")
bound_dag_univ = dagmc_univ.bounded_universe()  
for cell in bound_dag_univ.get_all_cells().values():
    region = cell.region
    if region:
        for halfspace in region:
            surface = halfspace.surface
            surface.boundary_type = "reflective"`
May I ask whether this method is feasible?
Thanks!

**Reply by aprilnovak** [reply]
Hi @Kevrog1006 , first let me apologize for the extremely long time which has passed since your question. I think I missed the initial email and am seeing this now because someone else message me with a link to your question. Apologies!

From within Cubit, you can export a DAGMC geometry if you start from either of the following meshes: (i) a surface mesh or (ii) a volume mesh. Does that answer your first question?

Regarding your second question, I _think_ that DAGMC does not support reflective or periodic boundary conditions - only vacuum. @pshriwise can confirm if that's the case.

**Reply by pshriwise** [reply]
Hi @Kevrog1006. Echoing what April said, apologies for the delay. 

When combining DAGMC geometry with CSG as you are above, the reflecting boundary conditions on the CSG surfaces should work just fine.

You can set reflecting boundary conditions on DAGMC surfaces directly as well, but given how much easier it's gotten to get a bounding CSG universe for the DAGMC model, I would only do that if necessary these days.

Hope this helps!!

**Reply by Kevrog1006** [reply]
It doesn't matter! Thank you both for your help @aprilnovak @pshriwise  .
I've made new progress on these issues. Firstly, I directly use the volume mesh in Cubit and there's no need to use the surface mesh. I believe that the volume mesh can replace the surface mesh. Currently, due to the performance limitations of my personal computer, I can only create a relatively rough unstructured mesh. However, in my tetrahedral volume mesh, there are more than one holes in the geometry, which may affect the calculation of the flow state and heat transfer. I think what is affected is only the shape of my wall surface, and this problem can be solved by refining the mesh, so the problem has been solved. Secondly, I used Cubit to draw the region of water outside the geometry and established the total reflection boundary condition with a toroidal CSG structure. Specifically, the inner diameter of my toroidal region is slightly larger than the maximum diameter of my fuel rod and smaller than my cylindrical water region. The outer diameter of the toroidal region can be equal to the radius of the cylindrical water region or slightly larger, and now I can perform normal calculations to obtain results.I think it is consistent with the second method mentioned @pshriwise . 
Thank you very much for your enthusiastic reply!
